rfcs/20201221-tfmot-compression-api.md

```diff
@@ -0,0 +1,447 @@
+     tool such as zip to the model file).
+
+     This interface is a purely functional one.
+  """
+
+  @abc.abstractmethod
+  def get_compressible_weights(
+      self, original_layer: tf.keras.layers.Layer) -> List[str]:
+    """Define compressible weights for each layer.
+
+    Args:
+       original_layer: tf.keras.layers.Layer representing a layer from the
+       original model.
+
+    Returns:
+       List of attribute names as string representing list of compressible
+       weights for the given layer. (e.g. return value ['kernel'] means
+       layer.kernel is compressible.)
+    """
+
+  @abc.abstractmethod
+  def init_training_weights(
+      self, pretrained_weight: tf.Tensor):
+    """Initialize training weights for the training model. It calls the `add_training_weight` method several times to add training weights.
+
+    Args:
+      pretrained_weight: tf.Tensor of a pretrained weight of a layer that will
+        be compressed eventually.
+    """
+
+  def add_training_weight(
+      self, *args, **kwargs):
+    """Add training weight for the training model. This method is called from `init_training_weights`.
+
+    Args:
+      *args, **kwargs: args and kwargs for training_model.add_weight.
+    """
+
+  @abc.abstractmethod
+  def project_training_weights(self, *training_weights: tf.Tensor) -> tf.Tensor:
+    """Define a piece of the forward pass during training, which operates on a single compressible weight.
+    The default throws an error when training occurs.
+
+    Args:
+       *training_weights: tf.Tensors representing any variables used during
+         training, for a single compressible weight, in the order returned in
+         `init_training_weights`.
+
+    Returns:
+       tf.Tensor to set the compressible weight to.
+    """
+
+  def update_training_weight(self, index: integer, tensor: tf.Tensor):
+    """Update a training weight on an index to a given tensor value.
+
+    This method is for the case that training weight should update to specific
+    value not from the model optimizer. It'll throws an error if it can't
+    find the training weight.
+
+    Args:
+      index: integer indicates index of training weight to update.
+      tensor: tf.Tensor to update specific training weight.
+    """
+
+  @abc.abstractmethod
+  def compress_training_weights(self, *training_weights: tf.Tensor) -> List[tf.Tensor]:
+    """Define the operations to compress a single weight’s training form after training.
+
+    'compress_training_weights' can refer to making the weight more amenable to compression
+    or actually compress the weight.
+
+    The default is an identity.
+
+    Args:
+      *training_weights: tf.Tensors representing all variables used during
+        training, for a single compressible weight, in the order returned in
+        `init_training_weights`.
+
+    Returns:
+      List of tf.Tensors to set to compressed or more compressible form.
```

Comment:
In there a 1:1 mapping between input tensors and output tensors?

Reply:
No, Just a function that has n inputs (training_weights) and m outputs. (compressed_weights)
It's different for each compression algorithm.